lib/ecmascript.ts

```diff
@@ -2137,123 +2129,191 @@ function CalendarDateUntil(
   calendar: Temporal.CalendarProtocol,
   dateLike: CalendarProtocolParams['monthCode'][0]
 ) {
-  const result = calendar.monthCode(dateLike);
-  if (result === undefined) {
-    throw new RangeError('calendar monthCode result must be a string');
+  const monthCode = GetMethod(calendar, 'monthCode');
+  let result = Call(monthCode, calendar, [dateLike]);
+  if (typeof result !== 'string') {
+    throw new TypeError('calendar monthCode result must be a string');
   }
-  return ToString(result);
+  return result;
 }
 
-export function CalendarDay(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['day'][0]) {
-  const result = calendar.day(dateLike);
-  return ToPositiveIntegerWithTruncation(result);
+export function CalendarDay(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['era'][0]) {
+  const eraYear = GetMethod(calendar, 'day');
+  let result = Call(eraYear, calendar, [dateLike]);
+  if (typeof result !== 'number') {
+    throw new TypeError('calendar day result must be a positive integer');
+  }
+  if (!IsIntegralNumber(result) || result < 1) {
+    throw new RangeError('calendar day result must be a positive integer');
+  }
+  return result;
 }
 
 export function CalendarEra(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['era'][0]) {
-  let result = calendar.era(dateLike);
-  if (result !== undefined) {
-    result = ToString(result);
+  const era = GetMethod(calendar, 'era');
+  let result = Call(era, calendar, [dateLike]);
+  if (result === undefined) {
+    return result;
+  }
+  if (typeof result !== 'string') {
+    throw new TypeError('calendar era result must be a string or undefined');
   }
   return result;
 }
 
-export function CalendarEraYear(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['eraYear'][0]) {
-  let result = calendar.eraYear(dateLike);
-  if (result !== undefined) {
-    result = ToIntegerWithTruncation(result);
+export function CalendarEraYear(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['era'][0]) {
+  const eraYear = GetMethod(calendar, 'eraYear');
+  let result = Call(eraYear, calendar, [dateLike]);
+  if (result === undefined) {
+    return result;
+  }
+  if (typeof result !== 'number') {
+    throw new TypeError('calendar eraYear result must be an integer or undefined');
+  }
+  if (!IsIntegralNumber(result)) {
+    throw new RangeError('calendar eraYear result must be an integer or undefined');
   }
   return result;
 }
 
-export function CalendarDayOfWeek(
-  calendar: Temporal.CalendarProtocol,
-  dateLike: CalendarProtocolParams['dayOfWeek'][0]
-) {
-  return ToPositiveIntegerWithTruncation(calendar.dayOfWeek(dateLike));
+export function CalendarDayOfWeek(calendar: Temporal.CalendarProtocol, dateLike: CalendarProtocolParams['era'][0]) {
+  const dayOfWeek = GetMethod(calendar, 'dayOfWeek');
+  const result = Call(dayOfWeek, calendar, [dateLike]);
+  if (typeof result !== 'number') {
```

Comment:
There is so much repetition across all these calendar methods... can we de-dupe this?